doc/devel/MEP/MEP30.rst

```diff
@@ -0,0 +1,266 @@
+Additionally, the set of plotting functions in the ``pyplot``
+namespace is controlled by Matplotlib upstream and is somewhat
+limited.  There have been proposals to provide a 'register' hook [dig
+up PR from 2 years ago] into the namespace, the only way for
+third-party packages to add to the pyplot namespace is monkey
+patching.  Either way, this can be problematic if two packages try to
+use the same name and which one gets it depends on import order.  Uses
+can write functions that 'feel' like ``pyplot`` functions by using
+``pyplot.gca()`` in their code, however this then ties them very
+tightly to ``pyplot`` which makes it more difficult to re-use the
+functions in embedding applications where the user may not want to
+import ``pyplot`` at all.
+
+
+The ``OO`` API is a bit less concise to work with interactively, it
+requires the user to explicitly create an ``Axes`` object to work
+with, however it solves the problem of shared global state.  The
+``Axes`` objects serve as both nodes in the draw tree and as the
+primary namespace for plotting functions.  It is possible (and
+encouraged) for users to write functions that take in ``Axes`` /
+``Figure`` objects and internally use the ``OO`` API, however these
+feel qualitatively different than the 'native' plotting routines which
+are ``Axes`` methods.  This can be overcome by ex sub-classing the
+``Axes`` or monkey patching methods on to it, however this has the
+same problem of clashing third-party packages.
+
+Among third-party functions in the wild (ex ``seaborn``, ``pandas``,
+and user code) the return types vary between returning the artist
+created during the call, the ``Axes`` objects plotted to, custom
+types, and nothing.
+
+Matplotlib is being used in a applications that are well beyond the
+original use cases and well beyond the expertise of the current
+development team.  To that end, we should make sure it is easy to
+write plotting functions that feel "native".  This can then be used to
+support an eco-system of ``mplkit-*`` domain specific libraries.  This
+libraries will be able to depend on a wider range of libraries (ex
+``scipy``, various scikits, and ``pandas``) than core Matplotlib and can
+be built around the 'fundamental' data structures of the domain.
+
+To better support the wide range of use cases, in a way that feels
+'native', we propose the following changes:
+
+ 1. move away from ``Axes`` as the primary namespace for plotting
+    routines
+ 2. provide decorators to easily opt third-party code into the
+    ``pyplot`` state machine without much boiler plate.
+
+
+Signatures
+----------
+
+There are two obvious ways to write a function has a required as input
+``Axes`` in a way that can be easily wrapped by a decorator::
+
+  def plotting_func(ax, *data_args, **style_kwargs):
+      ...
+
+or ::
+
+  def ploting_func(*data_args, ax, **style_kwargs):
+      ...
+
+The first case has the advantage that it works in both python2 and
```

Comment:
hopefully py2 won't really be a thing anymore by the time this is implemented :p

Reply:
indeed!  If you dig into the commits on this, I started writing this in 2016...